system/Model.php

```diff
@@ -730,20 +730,17 @@ public function insert($data = null, bool $returnID = true)
 		// If insertion succeeded then save the insert ID
 		if ($result)
 		{
-			if ( ! $this->hasAutoincrement)
+			if (! $this->useAutoIncrement)
 			{
-				if (isset($data[$this->primaryKey]))
+				if (empty($data[$this->primaryKey]))
 				{
-					$this->insertID = $data[$this->primaryKey];
-				}
-				else
-				{
-					throw new \Exception('Can\'t return primary key');
+					throw DataException::forEmptyDataset('pk');
```

Comment:
I know I suggested it, but now I just checked the message behind this exception and I think we should create a new exception.
```suggestion
					throw DataException::forEmptyPrimaryKey();
```
This would require to create a new method in `system/Database/Exceptions/DataException.php` file and also an entry to the language file - something like "You must provide a Primary Key value".

Reply:
I think the primary key cannot always be transmitted and this should not be a mistake.

Now I don't know this case, in example - primary key is random value from database. We cant got it from LASTVAL() and it isn't in $data array.

Best way edit insert SQL string to
INSERT INTO table (v) VALUES('val') RETURNING {primary key}

Reply:
I don't know how to do it

Reply:
I understand your reasoning behind mentioning the `RETURNING` in the `INSERT` query, but this is more like a separate problem/idea that should be addressed in a different PR.

Here we are trying to deal only with the scenario where a developer is responsible for providing the primary key.